### PR TITLE
Support for ENUM types

### LINKED ISSUE #7626: sql: enable enum values to be stored in tables and used in queries
PR: https://github.com/cockroachdb/cockroach/pull/48556

From release notes:
> Added support for Postgres `ENUM` types to be used in queries and stored in tables. [#48556][#48556] {% comment %}doc{% endcomment %}

### LINKED ISSUE #7751: Support for ENUM and SET types
https://cockroachlabs.atlassian.net/browse/CRDB-143

### LINKED ISSUE #8136: sql: add support for alter type rename value
PR: https://github.com/cockroachdb/cockroach/pull/51490

From release notes:
> Added support for renaming a value of an `ENUM` using the `ALTER TYPE <type> RENAME VALUE <old> TO <new>` syntax. [#51490][#51490] {% comment %}doc{% endcomment %}

### LINKED ISSUE #8154: sql: new statement `SHOW ENUMS`
PR: https://github.com/cockroachdb/cockroach/pull/52538

From release notes:
> Added a new statement `SHOW ENUMS` which displays information on existing enums. [#52538][#52538] {% comment %}doc{% endcomment %}

### PR DESCRIPTION
**Pull request by ericharmeling**
Fixes #7751. 
Fixes #7626. 
Fixes #8154.
Fixes #8136.

### CODE REVIEW

File: v20.2/create-type.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+> CREATE TABLE accounts (
+        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+        balance DECIMAL,
+        status status
+);
+~~~
+
+{% include copy-clipboard.html %}
+~~~ sql
+> INSERT INTO accounts(balance,status) VALUES (500.50, 'open'),(0.00, 'closed'),(1.25, 'inactive');
```
Comment (rohany) [inline v20.2/create-type.md:56]: spaces in between the values rows

File: v20.2/create-type.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+~~~
+
+{% include copy-clipboard.html %}
+~~~ sql
+> SELECT * FROM accounts;
```
Comment (rohany) [inline v20.2/create-type.md:61]: maybe add an example where you use the enum in a filter

File: v20.2/data-types.md [hunk 3]
```diff
@@ -16,6 +16,7 @@ Type | Description | Example | [Vectorized Execution](vectorized-execution.html)
 [`BYTES`](bytes.html) | A string of binary characters. | `b'\141\061\142\062\143\063'` | Supported
 [`COLLATE`](collate.html) | The `COLLATE` feature lets you sort [`STRING`](string.html) values according to language- and country-specific rules, known as collations.  | `'a1b2c3' COLLATE en` | Not supported
 [`DATE`](date.html) | A date.  | `DATE '2016-01-25'` | Supported
+[`ENUM`](enum.html) | <span class="version-tag">New in v20.2:</span> A data type comprised of a set of static values. | `ENUM ('club, 'diamond', 'heart', 'spade')` | Not supported
```
Comment (rohany) [inline v20.2/data-types.md:19]: I would mention that it is user defined/user created.

It still says not supported here

File: v20.2/drop-type.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
+---
+title: DROP TYPE
+summary: The DROP TYPE statement drops an enumerated data type from the database.
+toc: true
+---
+
+<span class="version-tag">New in v20.2:</span> The `DROP TYPE` [statement](sql-statements.html) drops a specified [enumerated data type](enum.html) from the database.
+
+{{site.data.alerts.callout_info}}
+You cannot drop a type that is in use by a table.
```
Comment (rohany) [inline v20.2/drop-type.md:10]: Or view

File: v20.2/drop-type.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+{{site.data.alerts.end}}
+
+## Required privileges
+
+The user must have the `DROP` [privilege](authorization.html#assign-privileges) on the parent database.
```
Comment (rohany) [inline v20.2/drop-type.md:15]: cc @RichardJCai I'm not sure this is the case. Is it that the user must be the owner of the type?

File: v20.2/create-type.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+~~~
+   name  |        value
+---------+-----------------------
+  status | open|closed|inactive
+(1 row)
+~~~
+
+{% include copy-clipboard.html %}
+~~~ sql
+> CREATE TABLE accounts (
```
Comment (rohany) [inline v20.2/create-type.md:47]: cc @RichardJCai 

I think we want to mention that a user must have `USAGE` privilege in order to add it to a table like this

File: v20.2/enum.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+## Syntax
+
+To declare a new `ENUM` data type, use [`CREATE TYPE`](create-type.html):
+
+~~~ sql
+> CREATE TYPE <name> AS ENUM ('<value1>', '<value2>', ...);
+~~~
+
+where `<name>` is the name of the new type, and `<value1>, <value2>, ...` are string literals that make up the type's set of static values.
```
Comment (rohany) [inline v20.2/enum.md:17]: maybe note (and give an example) that `name` can be qualified with a db and schema

File: v20.2/enum.md [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+~~~
+
+
+## Supported casting and conversion
+
+`ENUM` data type values can be [cast](data-types.html#data-type-conversions-and-casts) to any of the following data types:
+
+Type | Details
+-----|--------
+`ARRAY` | Requires supported [`ARRAY`](array.html) string format, e.g., `'{1,2,3}'`.<br>Limited to `ARRAY`s of type [`INT`](int.html) and [`DECIMAL`](decimal.html).
```
Comment (rohany) [inline v20.2/enum.md:105]: I don't think that these casts are valid. How did you generate these?

File: v20.2/drop-type.md (hunk 5 above)
Comment (RichardJCai) [reply v20.2/drop-type.md:15]: Yep, owner or member of the owner role.

Also considering adding DROP privilege on the type itself, not sure yet.

**Comment by rohany** [comment]
> Do we have a list of valid casts? I know STRING is one of them.

Just string and enum

**Comment by lnhsingh** [comment]
FYI, just merged: https://github.com/cockroachdb/docs/pull/8052

Can you rebase and add links on the `v20.2/restore.html` page to `enum.html` and `drop-type.html`? Thank you!

File: v20.2/alter-type.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+<div>
+  {% include {{ page.version.version }}/sql/diagrams/alter_type.html %}
+</div>
+
+## Parameters
+
+Parameter | Description
+----------|------------
+`type_name` | The name of the user-defined type.
+`ADD VALUE` | Add a constant value to the user-defined type's list of values.
```
Comment (rohany) [inline v20.2/alter-type.md:20]: make sure the cover the `ADD VALUE {BEFORE | AFTER} 'val'` option. This adds a new value to the enum, but places its sort order relative to the target existing value.

File: v20.2/create-type.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+        balance DECIMAL,
+        status status
+);
+~~~
+
+{% include copy-clipboard.html %}
+~~~ sql
+> INSERT INTO accounts(balance,status) VALUES (500.50,'open'),(0.00,'closed'),(1.25,'inactive');
```
Comment (rohany) [inline v20.2/create-type.md:56]: [nit]: spaces between the value clause members

File: v20.2/create-type.md (hunk 6 above)
Comment (rohany) [reply v20.2/create-type.md:47]: bump

**Comment by rohany** [comment]
> @rohany USAGE on the type, or on the table? (Or on both?)

You need `USAGE` on a type to add it as a table column, or use it in a table's default expression or something. The intuition is that adding a type to a table creates a reference from the type to the table (you can't drop the type unless the table is dropped), so we want to require a user to have the permission to create such a dependency.

**Comment by rohany** [comment]
> I don't think so, but there's an issue to implement that feature: cockroachdb/cockroach#53566.

For now, you can get this information from the `pg_catalog.pg_type` system table.

**Pull request merged**